server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -709,11 +709,16 @@ public void onFailure(Exception e) {
 
         if (indexSettings.isSoftDeleteEnabled() && useRetentionLeasesInPeerRecovery == false) {
             final RetentionLeases retentionLeases = replicationTracker.getRetentionLeases();
+
             final Set<ShardRouting> shardRoutings = Sets.newHashSetWithExpectedSize(routingTable.size());
             for (int copy = 0; copy < routingTable.size(); copy++) {
-                shardRoutings.add(routingTable.shard(copy));
+                ShardRouting shardRouting = routingTable.shard(copy);
+                if (shardRouting.isPromotableToPrimary()) shardRoutings.add(shardRouting);
             }
-            shardRoutings.addAll(routingTable.assignedShards()); // include relocation targets
+
+            // include relocation targets
+            shardRoutings.addAll(routingTable.assignedShards().stream().filter(ShardRouting::isPromotableToPrimary).toList());
```

Comment:
Do we need to allocate an intermediate list here? In fact, this seems kinda weird, we already iterate over the routing table in the few lines above, let's just pull out the relocation targets in that loop instead of adding all the shards again.